cpp/src/pausable_thread_loop.cpp

```diff
@@ -12,8 +12,22 @@ PausableThreadLoop::PausableThreadLoop(std::function<void()> func, Duration slee
         while (true) {
             {
                 std::unique_lock<std::mutex> lock(mutex_);
-                cv_.wait(lock, [this]() { return !paused_ || !active_; });
-                if (!active_) {
+                cv_.wait(lock, [this]() { return state_ != State::Paused; });
+                // Note: state changes here must notify any waiters.
+                switch (state_) {
+                case State::Running:
+                    break;
+                case State::Pausing:
+                    state_ = State::Paused;
+                    lock.unlock();
+                    cv_.notify_one();
+                case State::Paused:
+                    continue;
+                case State::Stopping:
+                    state_ = State::Stopped;
+                    lock.unlock();
+                    cv_.notify_one();
```

Comment:
When we're in the intermediate stages, when we then flip the state we must notify any waiter.